File: java/src/main/java/com/experian/dvs/client/address/search/SearchResultFuture.java
package com.experian.dvs.client.address.search;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.RestApiResponseError;
import com.experian.dvs.client.server.address.search.RestApiAddressSearchResponse;

import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class SearchResultFuture implements Future<SearchResult> {

    private final Future<RestApiAddressSearchResponse> apiFuture;

    public SearchResultFuture(Future<RestApiAddressSearchResponse> apiFuture) {
        this.apiFuture = apiFuture;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiFuture.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiFuture.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiFuture.isDone();
    }

    @Override
    public SearchResult get() throws InterruptedException, ExecutionException {
        final RestApiAddressSearchResponse apiSearchResponse = this.apiFuture.get();
        final Optional<RestApiResponseError> optError = apiSearchResponse.getError();
        if (optError.isPresent()) {
          throw EDVSException.using(optError.get());
        }
        return new SearchResult(apiSearchResponse);
    }

    @Override
    public SearchResult get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiAddressSearchResponse apiSearchResponse = this.apiFuture.get(timeout, unit);
        final Optional<RestApiResponseError> optError = apiSearchResponse.getError();
        if (optError.isPresent()) {
            throw EDVSException.using(optError.get());
        }
        return new SearchResult(apiSearchResponse);
    }
}

File: csharp/DVSClient/DVSClient/address/search/SearchResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Address.Search;

namespace DVSClient.Address.Search
{
    public class SearchResultFuture
    {
        private readonly Task<RestApiAddressSearchResponse> _apiFuture;

        public SearchResultFuture(Task<RestApiAddressSearchResponse> apiFuture)
        {
            _apiFuture = apiFuture;
        }

        public async Task<SearchResult> GetAsync()
        {
            try
            {
                var apiSearchResponse = await _apiFuture;
                var error = apiSearchResponse.Error;
                if (error != null)
                {
                    throw EDVSException.Using(error);
                }
                return new SearchResult(apiSearchResponse);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<SearchResult> GetAsync(TimeSpan timeout)
        {
            try
            {
                var apiSearchResponse = await Task.WhenAny(_apiFuture, Task.Delay(timeout)) == _apiFuture
                    ? await _apiFuture
                    : throw new TimeoutException();
                var optError = apiSearchResponse.Error;
                if (optError != null)
                {
                    throw EDVSException.Using(optError);
                }
                return new SearchResult(apiSearchResponse);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiFuture.IsCompleted;
        }
    }
}